src/checkers/inference/solver/GeneralSolver.java

```diff
@@ -0,0 +1,411 @@
+package checkers.inference.solver;
+
+import java.lang.reflect.Constructor;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+import javax.annotation.processing.ProcessingEnvironment;
+import javax.lang.model.element.AnnotationMirror;
+
+import org.checkerframework.framework.type.QualifierHierarchy;
+import org.checkerframework.javacutil.ErrorReporter;
+
+import checkers.inference.DefaultInferenceSolution;
+import checkers.inference.InferenceSolution;
+import checkers.inference.InferenceSolver;
+import checkers.inference.model.ConstantSlot;
+import checkers.inference.model.Constraint;
+import checkers.inference.model.Serializer;
+import checkers.inference.model.Slot;
+import checkers.inference.model.VariableSlot;
+import checkers.inference.solver.backend.BackEnd;
+import checkers.inference.solver.constraintgraph.ConstraintGraph;
+import checkers.inference.solver.constraintgraph.GraphBuilder;
+import checkers.inference.solver.frontend.ConstraintSerializer;
+import checkers.inference.solver.frontend.Lattice;
+import checkers.inference.solver.frontend.TwoQualifiersLattice;
+import checkers.inference.solver.util.Constants;
+import checkers.inference.solver.util.PrintUtils;
+import checkers.inference.solver.util.StatisticRecorder;
+import checkers.inference.solver.util.StatisticRecorder.StatisticKey;
+
+/**
+ * GeneralSolver is the entry point of general solver framework, and it is also
+ * the front end of whole solver system. GeneralSolver configures command line
+ * arguments, creates corresponding back end(s) and serializer, invokes the back
+ * end(s) and returns the solution.
+ * 
+ * @author jianchu
+ *
+ */
+
+public class GeneralSolver implements InferenceSolver {
+
+    protected String backEndType;
+    protected boolean useGraph;
+    private boolean solveInParallel;
+    private boolean collectStatistic;
+    private Lattice lattice;
+    private ConstraintGraph constraintGraph;
+    private BackEnd<?, ?> realBackEnd;
+
+    // Timing variables:
+    private long graphBuildingStart;
+    private long graphBuildingEnd;
+    private long solvingStart;
+    private long solvingEnd;
```

Comment:
I think I can make the graph one local, but the solving time one is used in multiple methods.